src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9539,16 +9539,24 @@ private bool ScanDesignator()
 
         private bool IsPossibleLambdaExpression(Precedence precedence)
         {
-            if (precedence <= Precedence.Lambda && this.PeekToken(1).Kind == SyntaxKind.EqualsGreaterThanToken)
+            if (precedence <= Precedence.Lambda)
             {
-                return true;
+                if (this.PeekToken(1).Kind == SyntaxKind.EqualsGreaterThanToken)
+                {
+                    return true;
+                }
+                else if (this.PeekToken(1).Kind == SyntaxKind.ExclamationToken)
+                {
+                    if ((this.PeekToken(2).Kind == SyntaxKind.EqualsGreaterThanToken))
+                    {
+                        return true;
+                    }
+                }
```

Comment:
i personally think this would read better as:

```c#
// explanation comment:
var token1 = this.PeekToken(1);
if (token1.Kind == ...)
{
    return true;
}

// explanation comment:
if (token1.Kind == ... && this.PeekToken(2).Kind == ...)
{
     return true;
}
```

It's a nice linear sequence of documented cases explaining what each is for, with a simple check for each.  With the nesting here it pulls things apart and makes it harder to see the relations being checked for.

I relaize this is minimal given the simplicity here. However, it's these tyeps of patterns that scale up well, and would be great to have across teh parser to make it easier to understand and work with.

Reply:
@fayrose consider adding the comments Cyrus mentioned.

---
In reply to: [286138981](https://github.com/dotnet/roslyn/pull/35826#discussion_r286138981) [](ancestors = 286138981)